### PR TITLE
[JIT] Re-enable alias sensitive peepholes

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#65860 [JIT] Re-enable alias sensitive peepholes**

Re-enable peepholes like `x + 0 == x`. These were at one point enabled, and then disabled because they did not properly account for aliasing, and then re-enabled with reconstructing the alias db everytime which is slow  - O(n^2). I've added correctness conditions, and I've also made it so that we avoid using stale aliasing properties for either the input or output of nodes we optimize. 
Some of the other code that we have written to avoid re-instantiating the alias db involves internally mutating it, however this is tricky to reason about and we probably have to add some extra invariants... 

cc @navahgar relevant to graph opts and @d1jang alias analysis relevant here

Differential Revision: [D31352382](https://our.internmc.facebook.com/intern/diff/D31352382)

### CODE REVIEW

File: torch/csrc/jit/passes/peephole_alias_sensitive.cpp [hunk 1]
```diff
@@ -7,16 +7,14 @@
 #include <torch/csrc/jit/passes/peephole_alias_sensitive.h>
 #include <torch/csrc/jit/runtime/graph_executor.h>
 #include <torch/csrc/utils/memory.h>
+#include <unordered_set>
```
Comment (jjsjann123) [inline torch/csrc/jit/passes/peephole_alias_sensitive.cpp:10]: Do we need this header?

Comment (eellison) [reply torch/csrc/jit/passes/peephole_alias_sensitive.cpp:10]: VSCode always does this its so annoying.. im going to figure out how to turn it off

**Review by jjsjann123 — APPROVED** [review]
Tried this with a simple toy example and looks like we are safely removing no-op mul in the backwards of add!
:shipit:

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31352382).

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31352382).

**Pull request closed without merging**